academy/2-main-concepts/queries.md

```diff
@@ -9,34 +9,328 @@ tag: deep-dive
+
+</HighlightBox>
+
+## Application query handling
+
+When a query is received by the full-node after being relayed from the underlying consensus engine, it is handled within an environment that understands application-specific types and has a copy of the state. [`BaseApp`](./base-app.md) implements the ABCI [query](https://docs.cosmos.network/master/core/baseapp.html#query) function and handles gRPC queries. The query route is parsed and matched to the fully-qualified service method name of an existing service method (most likely in one of the modules). `BaseApp` then relays the request to the relevant module.
+
+Apart from gRPC routes, `BaseApp` also handles four different types of queries:
+
+* `app`
+* `store`
+* `p2p`
+* `custom`
+
+The first three types, `app`, `store`, and `p2p`, are purely application-level and thus directly handled by `BaseApp` or the stores. The custom query type requires `BaseApp` to route the query to a module's [legacy queriers](https://docs.cosmos.network/master/building-modules/query-services.html#legacy-queriers).
+
+<HighlightBox type="tip">
+
+To learn more about these queries, check out the [Cosmos SDK documentation on Tendermint RPC](https://docs.cosmos.network/master/core/grpc_rest.html#tendermint-rpc).
 
 </HighlightBox>
 
-A query is a request for information made by end-users of an application through an interface and processed by a full node. Available information includes:
+Since `MyQuery` has a fully-qualified Protobuf service method name from the staking module, `BaseApp`:
 
-* Information about the network
-* Information about the application itself
-* Information about the application state
+1. Parses the path.
+2. Retrieves the corresponding gRPC handler with its own internal `GRPCQueryRouter`.
+3. Routes the query to the module.
 
-Queries do not require consensus to be processed as they do not trigger state transitions. Therefore, queries can be fully handled independently by a full node.
+The gRPC handler is responsible for recognizing the query, retrieving the appropriate values from the application's stores, and returning a response.
 
-<HighlightBox info=”info”>
+<HighlightBox type="tip">
 
-Visit the [detailed Cosmos SDK documentation](https://docs.cosmos.network/master/basics/query-lifecycle.html) to get a clear overview of the query lifecycle and learn how a query is created, handled, and responded to.
+Take a closer look at the [Cosmos documentation on query services](https://docs.cosmos.network/master/building-modules/query-services.html) for more on query services.
 
 </HighlightBox>
 
+Once a result is received from the querier, `BaseApp` begins the process of returning a response to the user.
+
+## Response
+
+`BaseApp` returns the response as an [`abci.ResponseQuery`](https://tendermint.com/docs/spec/abci/abci.html#messages) type since `Query()` is an ABCI function. The `client.Context` `Query()` routine receives the response and responds to the user.
```

Comment:
```suggestion
`BaseApp` returns the response as an [`abci.ResponseQuery`](https://docs.tendermint.com/master/spec/abci/abci.html#messages) type since `Query()` is an ABCI function. The `client.Context` `Query()` routine receives the response and responds to the user.
```